qiling/os/uefi/smst.py

```diff
@@ -0,0 +1,325 @@
+
+	return EFI_SUCCESS
+
+@dxeapi(params = {
+	"Memory"	: ULONGLONG,	# EFI_PHYSICAL_ADDRESS
+	"Pages"		: ULONGLONG		# UINTN
+})
+def hook_SmmFreePages(ql, address, params):
+	address = params["Memory"]
+	alloc_size = params["Pages"] * PAGE_SIZE
+
+	ret = ql.mem.free(address, alloc_size)
+
+	return EFI_SUCCESS if ret else EFI_INVALID_PARAMETER
+
+@dxeapi(params = {
+	"PoolType"	: INT,		# EFI_MEMORY_TYPE
+	"Size"		: INT,		# UINTN
+	"Buffer"	: POINTER	# PTR(PTR(VOID))
+})
+def hook_SmmAllocatePool(ql, address, params):
+	# TODO: allocate memory acording to "PoolType"
+	address = ql.loader.smm_context.heap.alloc(params["Size"])
+	write_int64(ql, params["Buffer"], address)
+
+	return EFI_SUCCESS if address else EFI_OUT_OF_RESOURCES
+
+@dxeapi(params = {
+	"Buffer": POINTER # PTR(VOID)
+})
+def hook_SmmFreePool(ql, address, params):
+	address = params["Buffer"]
+	ret = ql.loader.smm_context.heap.free(address)
+
+	return EFI_SUCCESS if ret else EFI_INVALID_PARAMETER
+
+@dxeapi(params = {
+	"Procedure"		: POINTER,
+	"CpuNumber"		: INT,
+	"ProcArguments"	: POINTER
+})
+def hook_SmmStartupThisAp(ql, address, params):
+	return EFI_INVALID_PARAMETER
+
+@dxeapi(params = {
+	"Handle"		: POINTER,		# PTR(EFI_HANDLE)
+	"Protocol"		: GUID,			# PTR(EFI_GUID)
+	"InterfaceType"	: ULONGLONG,	# EFI_INTERFACE_TYPE
+	"Interface"		: POINTER,		# PTR(VOID)
+})
+def hook_SmmInstallProtocolInterface(ql, address, params):
+	handle = read_int64(ql, params["Handle"])
+
+	if handle == 0:
+		handle = ql.loader.smm_context.heap.alloc(1)
+
+	dic = ql.loader.smm_context.protocols.get(handle, {})
+
+	dic[params["Protocol"]] = params["Interface"]
+	ql.loader.smm_context.protocols[handle] = dic
+	check_and_notify_protocols(ql)
+	write_int64(ql, params["Handle"], handle)
+
+	return EFI_SUCCESS
+
+@dxeapi(params = {
+	"Handle"	: POINTER,	# EFI_HANDLE
+	"Protocol"	: GUID,		# PTR(EFI_GUID)
+	"Interface"	: POINTER	# PTR(VOID)
+})
+def hook_SmmUninstallProtocolInterface(ql, address, params):
+	handle = params["Handle"]
+
+	if handle not in ql.loader.smm_context.protocols:
+		return EFI_NOT_FOUND
+
+	dic = ql.loader.smm_context.protocols[handle]
+	protocol = params["Protocol"]
+
+	if protocol not in dic:
+		return EFI_NOT_FOUND
+
+	del dic[protocol]
+
+	return EFI_SUCCESS
+
+@dxeapi(params = {
+	"Handle"	: POINTER,	# EFI_HANDLE
+	"Protocol"	: GUID,		# PTR(EFI_GUID)
+	"Interface"	: POINTER	# PTR(PTR(VOID))
+})
+def hook_SmmHandleProtocol(ql, address, params):
+	handle = params["Handle"]
+	protocol = params["Protocol"]
+	interface = params['Interface']
+
+	hdict = ql.loader.smm_context.protocols
+
+	if handle in hdict and protocol in hdict[handle]:
+		write_int64(ql, interface, hdict[handle][protocol])
+
+		return EFI_SUCCESS
+
+	return EFI_NOT_FOUND
+
+@dxeapi(params = {
+	"Protocol"		: GUID,		# PTR(EFI_GUID)
+	"Event"			: POINTER,	# EFI_EVENT
+	"Registration"	: POINTER	# PTR(PTR(VOID))
+})
+def hook_SmmRegisterProtocolNotify(ql, address, params):
+	event = params['Event']
```

Comment:
I've just found a bug (my bug), `SmmRegisterProtocolNotify` takes a function pointer, not an event like `RegisterProtocolNotify`. This will be a good opportunity to fix it if you can. Thank you! 
`EFI_STATUS
EFIAPI
SmmRegisterProtocolNotify (
  IN  CONST EFI_GUID              *Protocol,
  IN  EFI_SMM_NOTIFY_FN           Function,
  OUT VOID                        **Registration
  );`

Reply:
Yes, looks like a copy & paste error from its BS counterpart.
It looks like it would require implementing an additional notification list, but I must admit I haven't fully understood the existing mechanism. I applied a temp fix for now, and commented out the existing code.

Reply:
I will wait for the commit :) before i merge